hugo/content/docs/ast-types.md

```diff
@@ -3,4 +3,340 @@ title: "AST Type Inference"
+2. `+=` for assigning **multiple values** to a property, resulting in the property's type to be an array of the right side of the assignment.
+3. `?=` for assigning a **boolean** to a property, resulting in the property's type to be a `boolean`.
+
+```
+X: name=ID numbers+=INT (numbers+=INT)* isValid?='valid'?;
+
+// generates:
+interface X extends AstNode {
+    name: string
+    numbers: Array<number>
+    isValid: boolean
+}
+```
+
+The right side of an assignment can be any of the following:
+* A terminal rule or a data type rule, which results in the type of the property to be a built-in type.
+* A parser rule, which results is the type of the property to be the type of the parser rule.
+* A cross-reference, which results in the type of the property to be a *Reference* to the type of the cross-reference.
+* An alternative, which results in the type of the property to be a type union of all the types in the alternative.
+
+```
+X: 'x' name=ID;
+
+Y: crossValue=[X:ID] alt=(INT | X | [X:ID]);
+
+// generates:
+interface X extends AstNode {
+    name: string
+}
+
+interface Y extends AstNode {
+    crossValue: Reference<X>
+    alt: number | X | Reference<X>
+}
+```
+
+### Unassigned Rule Calls
+A parser rule does not necessarily need to have assignments. It may also contain only *unassigned rule calls*. These kind of rules can be used to change the types' hierarchy.
+```
+X: A | B;
+
+A: 'A' name=ID;
+B: 'B' name=ID count=INT;
+
+// generates:
+type X = A | B;
+
+interface A extends AstNode {
+    name: string
+}
+
+interface B extends AstNode {
+    name: string
+    count: number
+}
+```
+
+### Simple Actions
+Actions can be used to infer the type of the AST node **inside** of a parser rule. They can be viewed as *syntactic sugar* and can increase readability of the grammar.
+```
+X: 
+    {infer A} 'A' name=ID 
+  | {infer B} 'B' name=ID count=INT;
+
+// is equivalent to:
+X: A | B;
+A: 'A' name=ID;
+B: 'B' name=ID count=INT;
+
+// generates:
+type X = A | B;
+
+interface A extends AstNode {
+    name: string
+}
+
+interface B extends AstNode {
+    name: string
+    count: number
+}
+```
+
+### Assigned actions
+Actions can also be used to control the structure of AST node types. This is a more advanced topic, so we recommend getting familiar with the rest of the documentation before diving into this section.
+
+Let's consider two different grammars derived from the [Arithmetics example](https://github.com/langium/langium/blob/main/examples/arithmetics/src/language-server/arithmetics.langium). These grammars are designed to parse a document containing a single definition comprised of a name and an expression assignment, with an expression being an indefinite length of additions or a numerical value.
+
+The first one does not use assigned actions:
+```
+Definition: 
+    'def' name=ID ':' expr=Expression;
+
+Expression:
+    '(' Addition ')' | value=NUMBER;
+
+Addition infers Expression:
+    left=Expression ('+' right=Expression)*;
```

Comment:
I believe this grammar doesn't explain the issue well enough, since it's quite restrictive (it needs parenthesis to parse) and does some weird stuff with it's addition rule: When repeatedly parsing `right=Expression`, we override the previous `right` assignment, effectively "deleting" the previous value. I would recommend this grammar instead:

```suggestion
Definition: 
    'def' name=ID ':' expr=Expression;

Expression:
    Addition;

Addition infers Expression:
    left=Value ('+' right=Expression)?;
    
Primary infers Expression:
    '(' Expression ')' | {Literal} value=NUMBER;
```